### PR TITLE
AK: Add a new facility called TaggedPtr

### PR DESCRIPTION
**Pull request by junaire**
TaggedPtr is a pointer with additional data associated with it, which is commonly used when we want to save some memory.

This implementation uses LLVM's `PointerIntPair` as a reference and with some simplification.

### CODE REVIEW

**Comment by awesomekling** [comment]
Where is this meant to be used? We generally don't add things without having a specific use case in mind. :)

**Comment by junaire** [comment]
> Where is this meant to be used? We generally don't add things without having a specific use case in mind. :)

Hi Andreas, thanks for the early reply :) I think this facility can be used in the `LibJS` or other cases when you want to save some memory (store the extra flag or enum in the lowest bits of a pointer)

**Comment by junaire** [comment]
For example, in https://github.com/SerenityOS/serenity/blob/master/Userland/Libraries/LibJS/AST.h#L718
We can convert `m_argument (Expression)` into `TaggedPtr` and store `m_is_yield_from` in it. Then we wrapped the `TaggedPtr` to a `RefPtr`. By doing so, we can avoid wasting the 1-byte space (maybe with more padding?)

File: AK/TaggedPtr.h [hunk 1]
```diff
@@ -0,0 +1,90 @@
...
+namespace AK {
+
+namespace detail {
+constexpr unsigned log2(unsigned bits)
+{
+    if (bits == 1)
+        return 0;
+
+    return log2(bits / 2) + 1;
+}
```
Comment (Hendiadyoin1) [inline AK/TaggedPtr.h:21]: I am pretty sure we have that in `AK/IntegralMath`

File: AK/TaggedPtr.h [hunk 2]
```diff
@@ -0,0 +1,90 @@
...
+}
+
+template<typename PtrType, unsigned Bits, typename TagType = unsigned>
+class TaggedPtr {
+    intptr_t value = 0;
```
Comment (Hendiadyoin1) [inline AK/TaggedPtr.h:26]: private members at the end, like in the rest of the code base

Comment (Hendiadyoin1) [inline AK/TaggedPtr.h:26]: Also prefer `FlatPtr`

**Review by Hendiadyoin1 — COMMENTED** [review]
In addition to some nit picks,
As Kling said, without a use case in sight, this is unlikely to get merged

**Comment by junaire** [comment]
> In addition to some nit picks, As Kling said, without a use case in sight, this is unlikely to get merged

OK, I see. So I should add a usage somewhere, right?

File: AK/TaggedPtr.h (hunk 2 above)
Comment (junaire) [reply AK/TaggedPtr.h]: Not sure about this, but open to discussion :)

**Comment by MacDue** [comment]
> OK, I see. So I should add a usage somewhere, right?

Yes, generally unused code is not merged. But if you find a use in a few places that makes a meaningful improvement it is much more likely :^)

**Comment by junaire** [comment]
> For example, in https://github.com/SerenityOS/serenity/blob/master/Userland/Libraries/LibJS/AST.h#L718 We can convert `m_argument (Expression)` into `TaggedPtr` and store `m_is_yield_from` in it. Then we wrapped the `TaggedPtr` to a `RefPtr`. By doing so, we can avoid wasting the 1-byte space (maybe with more padding?)

I just realize that this stuff is not perfectly fit our situation :(

So the problem is that we rarely use raw pointers but use `RefPtr` and `NonullRefPtr` almost everywhere. And the type stored in `*RefPtr` must inherit from `RefCountBase`, so wrapping the real type in `TaggedPtr` doesn't compile. 

I'm discovering ways to solve it :)

**Pull request closed without merging**